### ISSUE TITLE
Meta Data section in the scenario JSON

### ISSUE THREAD

**Issue by authorjapps** [authorjapps/zerocode#207]
### Objective
+ This will help to capture any comments or useful information about authors or ticket(jira story)

The section could look like below-
```
{
    "scenarioName": "Scenario Verify Person REST apis as well as DB layer",
    "ignoreStepFailures": false, //default is false
    "steps": [
        {
            "name": "get_enitity",
            "url": "/wrong/url",
            "operation": "GET",
            "request": {
            },
            "assertions": {
                "status": 200
            }
        },
        "metaData":{
            "author" : "Emma",
            "ticket" : "ISSUE-519",
            "testPack" : "sit2"
            // and so on...
        }
    ]
}

```

Also the runners can be twigged by the user or the framework to process these fields.
e.g.
+ Show these in the reports etc.

**Comment by aghorpade** [comment]
@authorjapps  - metdata tag is associated with steps or should it be associated with scenario? I think if we associate metadata with scenario then it will have more significance and useful too in reporting.
E.g. Scenario - always comes with specific ticket and steps can have more than one steps so I think metadata should along with scenario.
Suggestions?

**Comment by authorjapps** [comment]
@aghorpade , Yes, you are exactly right 👍 . Updated the description. That was a typo

**Comment by jneate** [comment]
@authorjapps - Do you still require help with this issue? It's my first time contributing so I'll take your judgement about whether you feel I should tackle another issue first.

**Comment by authorjapps** [comment]
Hello @jneate , thanks for your interest in contributing.
Seems like we will have to expand this issue a bit, will add more details. You can then pick it.

In the mean time(if you get chance)
You can look at this bug => https://github.com/authorjapps/zerocode/issues/241
_(This was definitely working earlier, but introduced in the recent releases... ie couple versions ago)._

> This is a minor bug - Somewhere the report builder is missing an entry.

Hope this will give a good start for the first issue !
👍

**Comment by tanvimehta** [comment]
@authorjapps Hello! Do you still need help fixing this issue? If so, I'd love to give it a shot. Thanks!

**Comment by nirmalchandra** [comment]
Hello @tanvimehta , will refine the ACs slightly and update you soon. Thanks for your interest ✌️

**Comment by priya1712d** [comment]
@authorjapps Hello team! Do you still need help with this issue? I am starting my open contribution journey, can jump in for other issues as well! Thank you in advance :)

**Comment by chaitanya1607** [comment]
@authorjapps I am new to open source contribution, I would like to explore this project. Can you help me set up this project in my local. Please share if there is any documentation. I checked the readme.md file but was not able to figure it out. Thank you in advance :)

**Comment by authorjapps** [comment]
> @authorjapps Hello Team, I am new to open source contribution, I would like to explore this project. Can you help me set up this project in my local. Please share if there is any documentation. I checked the readme.md file but was not able to figure it out. Thank you in advance :)

@chaitanya1607, Thanks for your interest. Can you please use this page below to build and setup the project?
https://github.com/authorjapps/zerocode/blob/master/BUILDING.md

**Comment by DohaRamadan** [comment]
> > @authorjapps Hello Team, I am new to open source contribution, I would like to explore this project. Can you help me set up this project in my local. Please share if there is any documentation. I checked the readme.md file but was not able to figure it out. Thank you in advance :)
> 
> @chaitanya1607, Thanks for your interest. Can you please use this page below to build and setup the project? https://github.com/authorjapps/zerocode/blob/master/BUILDING.md

Can I contribute to this? I'm new here and looking for something I can do.

**Comment by prayascoriolis** [comment]
Hi @authorjapps 

I've thoroughly reviewed the Zerocode source code and have gained a good understanding of its functionality. I am eager to contribute to this issue and believe I have a conceptual approach to solve it. However, I have a few questions to clarify:

Where does the metadata come from? Will it be present in the scenario JSON files, for example, in http-testing/src/test/resources/helloworld/hello_world_status_ok_assertions.json?
Looking forward to your guidance.

Thank you!

**Comment by prayascoriolis** [comment]
I've figure out the folowing steps to implement this feature:

1. Update the ScenarioSpec class: Add a new field to hold the metadata.
2. Modify the ZeroCodeMultiStepsScenarioRunnerImpl class: Update the runScenario method to include metadata in the report.
3. Update the ZeroCodeExecReportBuilder class: Add a field for metadata and update the build method.
4. Update the ZeroCodeExecReport class: Add a field for metadata.
5. Modify the report generation: Update the ZeroCodeReportGeneratorImpl class to include metadata in the generated reports.
6. In the ZeroCodeUnitRunner class: Ensure that the metadata is properly passed from the JSON test case to the ScenarioSpec object.

Let me know if there's anything else i need to change to accomodate this feature?

**Comment by authorjapps** [comment]
@prayascoriolis , apologies for the delay in responding.
Thanks for picking this and trying to solutionize it.
Can you generate dummy/sample report and 
- attach the HTML file as a zip 
- attach few screenshots 
to get an idea or a preview of, how it's going to look to the end users?

Appreciate your effort!

**Comment by prayascoriolis** [comment]
> @prayascoriolis , apologies for the delay in responding. Thanks for picking this and trying to solutionize it. Can you generate dummy/sample report and
> 
> * attach the HTML file as a zip
> * attach few screenshots
>   to get an idea or a preview of, how it's going to look to the end users after you have implemented?
> 
> Appreciate your effort!

Thank you for your patience and for the opportunity to work on this feature. I've implemented the meta data section in the scenario JSON and updated the report generation process. Though it's still ambigous on how to get html reports (I can only get json reports from target dir), I've created a sample report to demonstrate how it will look for end users.

1. (present in src/test/java/org/jsmart/zerocode/testhelp/tests/MetaDataTest/MetaDataTest.java)
2. MetaDataTest.json (present in src/test/resources/MetaDataTest/MetaDataTest.json)
3. Metadataf30d7ce9-6f1f-44fc-be1f-0a239c61d445.json (present in target/zerocode-test-reports/Scenario - Get GitHub User Details with Metadataf30d7ce9-6f1f-44fc-be1f-0a239c61d445.json) 
[reports.zip](https://github.com/user-attachments/files/16691090/reports.zip)

**Comment by prayascoriolis** [comment]
> @prayascoriolis , apologies for the delay in responding. Thanks for picking this and trying to solutionize it. Can you generate dummy/sample report and
> 
> * attach the HTML file as a zip
> * attach few screenshots
>   to get an idea or a preview of, how it's going to look to the end users after you have implemented?
> 
> Appreciate your effort!

I f you're talking about zerocode-junit-interactive-fuzzy-search.html, this is the response i'm getting

![image](https://github.com/user-attachments/assets/70dee0fb-0ad5-4f38-9569-2ffb27d6b5e9)

**Comment by prayascoriolis** [comment]
@authorjapps, I've also attached the final csv ouput, sorry for sharing it late as there was a bug in including meta data in csv reports.

You can see the new added column name including: metaAuthors, metaTickets, metaCategories and metaOthers, with their corressponding values filled in last 10 rows

[zerocode-junit-granular-report.csv](https://github.com/user-attachments/files/16727666/zerocode-junit-granular-report.csv)

**Comment by prayascoriolis** [comment]
Hi @authorjapps ,

Have you had a chance to review the sample reports above? It would be great to revisit this issue whenever it's convenient for you.

Thanks for your time!

**Comment by authorjapps** [comment]
> Hi @authorjapps ,
> 
> Have you had a chance to review the sample reports above? It would be great to revisit this issue whenever it's convenient for you.
> 
> Thanks for your time!

Sorry @prayascoriolis , couldn't get chance to look at it earlier.
🟢 "pass" scenario looking good. 👍 
 
🔴 Can you attach a screenshot of the report how it looks when the scenario fails? 

That's because, failed ones are important to look at, the "meta" section should help in that aspect, then it would be useful.

**Comment by nirmalchandra** [comment]
Also, @prayascoriolis , can you attach the PR(draft version if you've any) to this ticket?
_(Do not worry even if it's not final one)_

**Comment by prayascoriolis** [comment]
Sure, @nirmalchandra @authorjapps, i've attached a zip file for reports where the scenario fails:
[metadata_in _failed_test.zip](https://github.com/user-attachments/files/17092936/metadata_in._failed_test.zip)

The required PR for this issue is here: https://github.com/authorjapps/zerocode/pull/674

**Comment by Aryan-Rustagi** [comment]
Hi @authorjapps, I'd like to work on this issue. I've read the discussion and the contribution guidelines. Could you please assign it to me?

**Comment by authorjapps** [comment]
> Hi [@authorjapps](https://github.com/authorjapps), I'd like to work on this issue. I've read the discussion and the contribution guidelines. Could you please assign it to me?

Hello @Aryan-Rustagi , 

Looks like a PR [here](https://github.com/authorjapps/zerocode/pull/674) is already merge with this feature. The ticket status was still open, hence confusion.

But what it's missing is an Integration Test i.e. a JSON scenario. 

Can you add and raise a new PR ? 

Also attach the report(download from Github Actions) from CI tab resulting from your PR please...
Then we can close this ticket.


---

Here is the code for "meta" FYI.

```java
    // Add getter and setter for meta
    public static Map<String, List<String>> getMeta() {
        return meta;
    }
    public void setMeta(Map<String, List<String>> meta) {
        this.meta = meta;
    }

    @JsonCreator
    public ScenarioSpec(
             ...
            @JsonProperty("meta") Map<String, List<String>> meta) {

```
